File: research/drift_env/simulation.py
"""Headless encounter core + safety guard — port of src/simulation.js."""

import math
from .physics import Physics
from .flight import FlightModel


class Simulation:
    def __init__(self, scenario=None, policy=None, decisionHz=1, flightHz=8, powerAuthority=1.0,
                 dSafe=0.5, hardBody=0.05, dvBudget=12.0, pcThreshold=1e-4, thrustAccel=0.0008, shield=True):
        self.scenario = scenario
        self.policy = policy
        self.decisionHz = decisionHz
        self.flightDt = 1.0 / flightHz
        self.powerAuthority = powerAuthority
        self.shield = shield  # A2: deterministic safety layer (validate + clamp + budget)
        self.phys = Physics(dSafe=dSafe, hardBody=hardBody, dvBudget=dvBudget, pcThreshold=pcThreshold)
        self.fm = FlightModel(self.phys, thrustAccel=thrustAccel)
        self.fm.set_power_authority(powerAuthority)
        self.log = []
        self.violations = 0  # actions the guard had to reject/clamp
        self._decision_acc = 0.0
        if scenario:
            self.reset(scenario)

    def reset(self, scenario=None):
        if scenario:
            self.scenario = scenario
        self.phys.arm(self.scenario)
        self.phys.enforce_budget = self.shield
        self.fm.cancel_burn()
        self.fm.set_power_authority(self.powerAuthority)
        self.log = []
        self.violations = 0
        self._decision_acc = math.inf  # decide immediately on first step

    def observe(self):
        fc = self.phys.forecast()
        threats = fc["all"] if fc else []
        worst = None
        worst_req = 0.0
        for th in threats:
            if th["safe"]:
                req = 0.0
            else:
                req = th["requiredDv"]
                if th["pc"] > self.phys.pcThreshold:
                    req = max(req, self.phys.required_dv_for_pc(th["target"]))
            if not th["safe"] and req >= worst_req:
                worst = th
                worst_req = req
        # can't request more Δv than is left (handles the unsolvable inf case)
        worst_req = min(worst_req, self.phys.dvRemaining)
        return {
            "worst": ({"target": worst["target"], "pc": worst["pc"], "predictedMiss": worst["predictedMiss"],
                       "tca": worst["tca"], "safe": worst["safe"]} if worst else None),
            "worstBurnDir": (self.phys.best_burn_dir(worst["target"]) if worst else {"x": 0, "y": 0, "z": 1}),
            "worstRecommendedDv": worst_req,
            "dSafe": self.phys.dSafe,
            "pcThreshold": self.phys.pcThreshold,
            "budgetRemaining": self.phys.dvRemaining,
            "dvUsed": self.phys.dvUsed,
            "parDv": self.phys.parDv,
            "powerAuthority": self.powerAuthority,
        }

    def execute(self, action):
        if not action or action.get("type") != "burn":
            self.fm.cancel_burn()
            return
        d = action.get("dir")
        raw = float(action.get("dv") or 0.0)

        if not self.shield:
            # No safety layer: pass the raw action straight through (A2 ablation).
            if d is None or raw <= 0:
                self.fm.cancel_burn()
                return
            self.fm.set_burn(d, raw)
            return

        # Deterministic safety guard: reject invalid, clamp to remaining budget.
        if not d or not all(math.isfinite(d.get(k, math.nan)) for k in "xyz"):
            self.violations += 1
            self.fm.cancel_burn()
            return
        dv = max(0.0, min(self.phys.dvRemaining, raw))
        if raw > self.phys.dvRemaining + 1e-9 or raw < 0:
            self.violations += 1  # policy asked for something out of bounds
        if dv <= 0:
            self.fm.cancel_burn()
            return
        self.fm.set_burn(d, dv)
        self.log.append({"t": round(self.phys.elapsed, 2), "dv": round(dv, 3)})

    def step(self, dt=None):
        dt = self.flightDt if dt is None else dt
        if self.phys.finished:
            return
        self._decision_acc += dt
        if self._decision_acc >= 1.0 / self.decisionHz:
            self._decision_acc = 0.0
            self.execute(self.policy["decide"](self.observe()))
        self.fm.tick(dt, "AGENT")
        self.phys.step(dt)

    def run(self, maxTime=60.0):
        t = 0.0
        while not self.phys.finished and t < maxTime:
            self.step(self.flightDt)
            t += self.flightDt
        p = self.phys
        return {
            "scenario": self.scenario.get("name"),
            "status": p.status,
            "resolved": p.status == "RESOLVED",
            "collision": p.status == "COLLISION",
            "dvUsed": p.dvUsed,
            "parDv": p.parDv,
            "ratio": (p.dvUsed / p.parDv) if p.parDv > 0 else None,
            "grade": p.grade,
            "minRangeAlly": p.minRangeAlly,
            "minRangeAsteroid": p.minRangeAsteroid,
            "actions": len(self.log),
            "violations": self.violations,
        }

File: src/simulation.js
// simulation.js — headless encounter core (Phase 0/2).
// Wraps Physics + FlightModel + a decision policy with NO DOM/Three/Reactor, so
// it runs identically in the browser and in Node (eval harness, tests, replays).
//
// It also implements the deterministic SAFETY GUARD: every action a policy
// proposes is validated and clamped here before it can touch the flight model.

import { Physics } from './physics.js';
import { FlightModel } from './flight_model.js';

export class Simulation {
  constructor(opts = {}) {
    this.scenario = opts.scenario;
    this.policy = opts.policy;
    this.decisionHz = opts.decisionHz ?? 1;
    this.flightDt = 1 / (opts.flightHz ?? 8);
    this.powerAuthority = opts.powerAuthority ?? 1;

    this.phys = new Physics({
      dSafe: opts.dSafe ?? 0.5,
      hardBody: opts.hardBody ?? 0.05,
      dvBudget: opts.dvBudget ?? 12,
      pcThreshold: opts.pcThreshold ?? 1e-4,
    });
    this.fm = new FlightModel(this.phys, { thrustAccel: opts.thrustAccel ?? 0.0008 });
    this.fm.setPowerAuthority(this.powerAuthority);

    this.log = [];            // action log (explainability)
    this._decisionAcc = 0;
    if (this.scenario) this.reset(this.scenario);
  }

  reset(scenario = this.scenario) {
    this.scenario = scenario;
    this.phys.arm(scenario);
    this.fm.cancelBurn();
    this.fm.setPowerAuthority(this.powerAuthority);
    this.log.length = 0;
    this._decisionAcc = Infinity; // decide immediately on first step
  }

  // Structured observation handed to the policy.
  // A threat is UNSAFE if Pc exceeds threshold OR it breaches the safe ring; the
  // recommended burn must satisfy BOTH (the literal mission + the Pc objective).
  observe() {
    const fc = this.phys.forecast();
    const threats = fc ? (fc.all || [fc.ally, fc.asteroid].filter(Boolean)) : [];
    let worst = null, worstReqDv = 0;
    for (const th of threats) {
      const reqDv = th.safe ? 0 : Math.max(th.requiredDv, th.pc > this.phys.pcThreshold ? this.phys.requiredDvForPc(th.target) : 0);
      if (!th.safe && reqDv >= worstReqDv) { worst = th; worstReqDv = reqDv; }
    }
    return {
      worst: worst && { target: worst.target, pc: worst.pc, predictedMiss: worst.predictedMiss, tca: worst.tca, safe: worst.safe },
      worstBurnDir: worst ? this.phys.bestBurnDir(worst.target) : { x: 0, y: 0, z: 1 },
      worstRecommendedDv: worstReqDv,
      dSafe: this.phys.dSafe,
      pcThreshold: this.phys.pcThreshold,
      budgetRemaining: this.phys.dvRemaining,
      dvUsed: this.phys.dvUsed,
      parDv: this.phys.parDv,
      powerAuthority: this.powerAuthority,
    };
  }

  // Deterministic safety guard: validate + clamp, then commit to the flight model.
  execute(action) {
    if (!action || action.type !== 'burn') { this.fm.cancelBurn(); return; }
    const d = action.dir;
    if (!d || !isFinite(d.x) || !isFinite(d.y) || !isFinite(d.z)) { this.fm.cancelBurn(); return; }
    const mag = Math.hypot(d.x, d.y, d.z);
    if (mag < 1e-9) { this.fm.cancelBurn(); return; }
    const dv = Math.max(0, Math.min(this.phys.dvRemaining, +action.dv || 0));
    if (dv <= 0) { this.fm.cancelBurn(); return; }
    this.fm.setBurn({ dir: d, targetDv: dv });
    this.log.push({ t: +this.phys.elapsed.toFixed(2), dir: { x: +d.x.toFixed(3), y: +d.y.toFixed(3), z: +d.z.toFixed(3) }, dv: +dv.toFixed(3) });
  }

  step(dt = this.flightDt) {
    if (this.phys.finished) return;
    this._decisionAcc += dt;
    if (this._decisionAcc >= 1 / this.decisionHz) {
      this._decisionAcc = 0;
      this.execute(this.policy.decide(this.observe()));
    }
    this.fm.tick(dt, 'AGENT');
    this.phys.step(dt);
  }

  run({ maxTime = 60 } = {}) {
    let t = 0;
    while (!this.phys.finished && t < maxTime) { this.step(this.flightDt); t += this.flightDt; }
    const p = this.phys;
    return {
      scenario: this.scenario.name,
      status: p.status,
      resolved: p.status === 'RESOLVED',
      collision: p.status === 'COLLISION',
      dvUsed: p.dvUsed,
      parDv: p.parDv,
      ratio: p.parDv > 0 ? p.dvUsed / p.parDv : null,
      grade: p.grade,
      minRangeAlly: p.minRangeAlly,
      minRangeAsteroid: p.minRangeAsteroid,
      actions: this.log.length,
    };
  }
}
